controller/service.go

```diff
@@ -180,3 +185,132 @@ func (c *controller) allocateIP(key string, svc *v1.Service) (net.IP, error) {
 	// Okay, in that case just bruteforce across all pools.
 	return c.ips.Allocate(key, isIPv6, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc))
 }
+
+// Dual-stack.
+
+func (c *controller) convergeBalancerDual(l log.Logger, key string, svc *v1.Service) bool {
+	var lbIP, lbIP2 net.IP
+
+	// The assigned LB IP is the end state of convergence. If there's
+	// none or a malformed one, nuke all controlled state so that we
+	// start converging from a clean slate.
+	if len(svc.Status.LoadBalancer.Ingress) > 1 {
+		lbIP = net.ParseIP(svc.Status.LoadBalancer.Ingress[0].IP)
+		lbIP2 = net.ParseIP(svc.Status.LoadBalancer.Ingress[1].IP)
+	}
+
+	// It's possible the config mutated and the IP we have no longer
+	// makes sense. If so, clear it out and give the rest of the logic
+	// a chance to allocate again.
+	if lbIP != nil && lbIP2 != nil {
+		// This assign is idempotent if the config is consistent,
+		// otherwise it'll fail and tell us why.
+		if err := c.ips.AssignDual(key, lbIP, lbIP2, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc)); err != nil {
+			l.Log("event", "clearAssignment", "reason", "notAllowedByConfig", "msg", "current IP not allowed by config, clearing")
+			c.clearServiceState(key, svc)
+			lbIP = nil
+		}
+
+		if lbIP != nil {
+			// The user might also have changed the pool annotation, and
+			// requested a different pool than the one that is currently
+			// allocated.
+			desiredPool := svc.Annotations["metallb.universe.tf/address-pool"]
+			if desiredPool != "" && c.ips.Pool(key) != desiredPool {
+				l.Log("event", "clearAssignment", "reason", "differentPoolRequested", "msg", "user requested a different pool than the one currently assigned")
+				c.clearServiceState(key, svc)
+				lbIP = nil
+			}
+		}
+	} else {
+		c.clearServiceState(key, svc)
+		lbIP = nil
+	}
+
+	// The (singular) svc.Spec.LoadBalancerIP is ignored for dual-stack
+	if svc.Spec.LoadBalancerIP != "" {
```

Comment:
Is this arbitrary? Given that we can't have two ips, here we are choosing to go against the user's will and assign two ips instead of the one he chose.

Reply:
We can reject of course. Is that preferable?